klayout/drc/testing/run_regression.py

```diff
@@ -0,0 +1,924 @@
+    df["in_rule_deck"] = df["in_rule_deck"].fillna(0)
+    df = df.merge(tc_df[["table_name", "run_status"]], how="left", on="table_name")
+
+    df["rule_status"] = "Passed"
+    df.loc[(df["false_negative"] > 0), "rule_status"] = "Rule Failed"
+    df.loc[(df["false_positive"] > 0), "rule_status"] = "Rule Failed"
+    df.loc[(df["pass_patterns"] < 1), "rule_status"] = "Rule Not Tested"
+    df.loc[(df["fail_patterns"] < 1), "rule_status"] = "Rule Not Tested"
+    df.loc[(df["in_rule_deck"] < 1), "rule_status"] = "Rule Not Implemented"
+    df.loc[~(df["run_status"].isin(["completed"])), "rule_status"] = "Test Case Run Failed"
+
+    return df
+
+
+def run_regression(drc_dir, output_path, target_table, cpu_count):
+    """
+    Running Regression Procedure.
+
+    This function runs the full regression on all test cases.
+
+    Parameters
+    ----------
+    drc_dir : string
+        Path string to the DRC directory where all the DRC files are located.
+    output_path : str
+        Path string to the location of the output results of the run.
+    target_table : string or None
+        Name of table that we want to run regression for. If None, run all found.
+    cpu_count : int
+        Number of cpus to use in running testcases.
+    Returns
+    -------
+    bool
+        If all regression passed, it returns true. If any of the rules failed it returns false.
+    """
+
+    ## Parse Existing Rules
+    rules_df = parse_existing_rules(drc_dir, output_path, target_table)
+    logging.info("## Total number of rules found in rule decks: {}".format(len(rules_df)))
+    logging.info("## Parsed Rules: \n" + str(rules_df))
+
+    ## Get all test cases available in the repo.
+    test_cases_path = os.path.join(drc_dir, "testing/testcases")
+    unit_test_cases_path = os.path.join(test_cases_path, "unit")
+    tc_df = build_tests_dataframe(unit_test_cases_path, target_table)
+    logging.info("## Total table gds files found: {}".format(len(tc_df)))
+    logging.info("## Found testcases: \n" + str(tc_df))
+
+    ## Run all test cases.
+    results_df, tc_df = run_all_test_cases(tc_df, drc_dir, output_path, cpu_count)
+    logging.info("## Testcases found results: \n" + str(results_df))
+    logging.info("## Updated testcases: \n" + str(tc_df))
+
+    ## Aggregate all dataframes into one
+    df = aggregate_results(tc_df, results_df, rules_df)
+    df.drop_duplicates(inplace=True)
+    logging.info("## Final analysis table: \n" + str(df))
+
+    ## Generate error if there are any missing info or fails.
+    df.to_csv(
+        os.path.join(output_path, "all_test_cases_results.csv"), index=False
+    )
+
+    ## Check if there any rules that generated false positive or false negative
+    failing_results = df[
+        ~df["rule_status"].isin(["Passed"])
+    ]
+    logging.info("## Failing test cases: \n" + str(failing_results))
+
+    if len(failing_results) > 0:
+        logging.error("## Some test cases failed .....")
+        return False
```

Comment:
+1 for returning structured result here, maybe we could keep the logging top level since we populate the result to the caller?

Reply:
Not sure what's the requirements here.

Reply:
Ideally, I think we should aim to keep most of the user facing logging (at least fatal `ERROR` and `INFO`) close to the top-level main.

Reply:
Done

Reply:
so did you decide to keep intermediate logging here (rather than bubbling it up to main?)

Reply:
@proppy All logging is brought to the main log.

Reply:
Yep, I understand that. The change I was suggesting here was too keep the user facing logging in the top level main function

Reply:
filed #23